source/common/jwt_authn/jwks.cc

```diff
@@ -0,0 +1,248 @@
+      return nullptr;
+    }
+    auto rsa =
+        bssl::UniquePtr<RSA>(RSA_public_key_from_bytes(castToUChar(pkey_der), pkey_der.length()));
+    if (!rsa) {
+      updateStatus(Status::JwksPemParseError);
+    }
+    return createEvpPkeyFromRsa(rsa.get());
+  }
+
+  bssl::UniquePtr<EVP_PKEY> createEvpPkeyFromJwkRSA(const std::string& n, const std::string& e) {
+    return createEvpPkeyFromRsa(createRsaFromJwk(n, e).get());
+  }
+
+  bssl::UniquePtr<EC_KEY> createEcKeyFromJwkEC(const std::string& x, const std::string& y) {
+    bssl::UniquePtr<EC_KEY> ec_key(EC_KEY_new_by_curve_name(NID_X9_62_prime256v1));
+    if (!ec_key) {
+      updateStatus(Status::JwksEcCreateKeyFail);
+      return nullptr;
+    }
+    bssl::UniquePtr<BIGNUM> bn_x = createBigNumFromBase64UrlString(x);
+    bssl::UniquePtr<BIGNUM> bn_y = createBigNumFromBase64UrlString(y);
+    if (!bn_x || !bn_y) {
+      // EC public key field is missing or has parse error.
+      updateStatus(Status::JwksEcParseError);
+      return nullptr;
+    }
+
+    if (EC_KEY_set_public_key_affine_coordinates(ec_key.get(), bn_x.get(), bn_y.get()) == 0) {
+      updateStatus(Status::JwksEcParseError);
+      return nullptr;
+    }
+    return ec_key;
+  }
+
+private:
+  bssl::UniquePtr<EVP_PKEY> createEvpPkeyFromRsa(RSA* rsa) {
+    if (!rsa) {
+      return nullptr;
+    }
+    bssl::UniquePtr<EVP_PKEY> key(EVP_PKEY_new());
+    EVP_PKEY_set1_RSA(key.get(), rsa);
+    return key;
+  }
+
+  bssl::UniquePtr<BIGNUM> createBigNumFromBase64UrlString(const std::string& s) {
+    std::string s_decoded = Base64Url::decode(s);
+    if (s_decoded == "") {
+      return nullptr;
+    }
+    return bssl::UniquePtr<BIGNUM>(BN_bin2bn(castToUChar(s_decoded), s_decoded.length(), NULL));
+  };
+
+  bssl::UniquePtr<RSA> createRsaFromJwk(const std::string& n, const std::string& e) {
+    bssl::UniquePtr<RSA> rsa(RSA_new());
+    rsa->n = createBigNumFromBase64UrlString(n).release();
+    rsa->e = createBigNumFromBase64UrlString(e).release();
+    if (!rsa->n || !rsa->e) {
+      // RSA public key field is missing or has parse error.
+      updateStatus(Status::JwksRsaParseError);
+      return nullptr;
+    }
+    return rsa;
+  }
+};
+
+} // namespace
+
+JwksPtr Jwks::createFrom(const std::string& pkey, Type type) {
+  JwksPtr keys(new Jwks());
+  switch (type) {
+  case Type::JWKS:
+    keys->createFromJwksCore(pkey);
+    break;
+  case Type::PEM:
+    keys->createFromPemCore(pkey);
+    break;
+  default:
+    PANIC("can not reach here");
+  }
+  return keys;
+}
+
+void Jwks::createFromPemCore(const std::string& pkey_pem) {
+  keys_.clear();
+  PubkeyPtr key_ptr(new Pubkey());
+  EvpPkeyGetter e;
+  key_ptr->evp_pkey_ = e.createEvpPkeyFromStr(pkey_pem);
+  key_ptr->pem_format_ = true;
+  updateStatus(e.getStatus());
+  if (e.getStatus() == Status::Ok) {
+    keys_.push_back(std::move(key_ptr));
+  }
+}
+
+void Jwks::createFromJwksCore(const std::string& pkey_jwks) {
+  keys_.clear();
```

Comment:
Not a fan of raw JSON parsing here and wondering if this could be cleaner.

JWKS is well-defined so may be you can define a proto with validate rules to handle them? Leaving a TODO is fine to me in this PR.

Reply:
Done